### PR TITLE
[DRAFT][onert-micro] Onert-micro training PoCv3

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This draft introduces third version of the onert-micro training runtime. This is for one-stage training (without generating backprop graph part) and without weight divider.

for issue: https://github.com/Samsung/ONE/issues/12873

ONE-DCO-1.0-Signed-off-by: Artem Balyshev [a.balyshev@samsung.com](mailto:a.balyshev@samsung.com)

### CODE REVIEW

File: onert-micro/eval-driver/TrainingDriver.cpp [hunk 1]
```diff
@@ -100,117 +92,236 @@ int entry(int argc, char **argv)
...
-  settings.metric = luci_interpreter::training::MAE;
-  onert_micro_training.test(num_test_data_samples, reinterpret_cast<const uint8_t *>(test_data),
-                            reinterpret_cast<const uint8_t *>(label_test_data),
-                            reinterpret_cast<void *>(&mae_result));
-
-  printf("MSE_ERROR TEST = %f\n", mse_result);
-
-  printf("MAE_ERROR TEST = %f\n", mae_result);
+  // Save training result
+  saveModel(output_trained_file_path, circle_model);
```
Comment (chunseoklee) [inline onert-micro/eval-driver/TrainingDriver.cpp:324]: This may be exposed via API. Thus, circle_model or model_ptr should be kept in some context.

File: onert-micro/eval-driver/TrainingDriver.cpp [hunk 2]
```diff
@@ -100,117 +92,236 @@ int entry(int argc, char **argv)
...
-    std::memcpy(input_data, train_data_u8, size);
+  // Create training interpreter and import models
+  onert_micro::OMTrainingInterpreter train_interpreter;
+  train_interpreter.importTrainModel(circle_model.data(), config);
 
-    interpreter.interpret();
-    auto data = reinterpret_cast<float *>(interpreter.readOutputTensor(0));
+  // Temporary buffer to read input data from file using BATCH_SIZE
+  float training_input[BATCH_SIZE * INPUT_SIZE];
+  float training_target[BATCH_SIZE * OUTPUT_SIZE];
```
Comment (chunseoklee) [inline onert-micro/eval-driver/TrainingDriver.cpp:226]: label data for CCE is one hot encoding. You will feed int label data by converting into float ?

File: onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+                                     forward_main_graph.getRuntimeContext(), b);
+
+    // Reset values
+    forward_main_graph.reset();
+  }
+
+  // Averaged result
+  {
+    float *f_metric_val = reinterpret_cast<float *>(metric_val);
+    *f_metric_val /= test_size;
```
Comment (Torrero) [inline onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:230]: Is it reasonable to insert assert for checking `test_size` equal zero?

File: onert-micro/eval-driver/TrainingDriver.cpp (hunk 1 above)
Comment (BalyshevArtem) [reply onert-micro/eval-driver/TrainingDriver.cpp:324]: So, move this function insert OMTrainingInterpreter, right?

File: onert-micro/eval-driver/TrainingDriver.cpp (hunk 2 above)
Comment (BalyshevArtem) [reply onert-micro/eval-driver/TrainingDriver.cpp:226]: Yes, I think it is one of the preprocessing step. And we have (for cross-etropy task) as input already preprocessed  one-hot encoded float data. Do you think it is better to move it into runtime?

File: onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp (hunk 3 above)
Comment (BalyshevArtem) [reply onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:230]: Yes, thank you, I will add assert here

File: onert-micro/onert-micro/src/train/metrics/MSE.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * E(Y, Y_t) = (SUM(Y_i - Y_t_i)^2) / N
+ * SUM - sum among i = (0 ... N - 1)
+ */
+float MSE::calculateValue(const uint32_t flat_size, float *calculated_data, float *target_data)
+{
+  float result_value = 0.f;
+
+  for (uint32_t i = 0; i < flat_size; ++i)
+  {
+    result_value += std::pow((calculated_data[i] - target_data[i]), 2);
```
Comment (Torrero) [inline onert-micro/onert-micro/src/train/metrics/MSE.cpp:35]: Here std::pow() can be replace by multiplication

**Comment by BalyshevArtem** [comment]
@chunseoklee, I added checkpoints saving and loading api:
https://github.com/Samsung/ONE/blob/ea703dd367b8f13a7a392a040e3f6029ce9c00bb/onert-micro/eval-driver/TrainingDriver.cpp#L318
https://github.com/Samsung/ONE/blob/ea703dd367b8f13a7a392a040e3f6029ce9c00bb/onert-micro/eval-driver/TrainingDriver.cpp#L326

It it similar to https://github.com/Samsung/ONE/issues/12997#issuecomment-2123904448

**Comment by BalyshevArtem** [comment]
> @chunseoklee, I added checkpoints saving and loading api:
> 
> https://github.com/Samsung/ONE/blob/ea703dd367b8f13a7a392a040e3f6029ce9c00bb/onert-micro/eval-driver/TrainingDriver.cpp#L318
> 
> https://github.com/Samsung/ONE/blob/ea703dd367b8f13a7a392a040e3f6029ce9c00bb/onert-micro/eval-driver/TrainingDriver.cpp#L326
> 
> It it similar to [#12997 (comment)](https://github.com/Samsung/ONE/issues/12997#issuecomment-2123904448)

To check format of the checkpoint file please see https://github.com/Samsung/ONE/discussions/13037#discussioncomment-9692320

File: onert-micro/eval-driver/TrainingDriver.cpp [hunk 5]
```diff
@@ -100,117 +92,241 @@ int entry(int argc, char **argv)
...
+    std::cout << "Run training for epoch: " << e + 1 << "/" << training_epochs << "\n";
+    config.training_context.num_epoch = e + 1;
+    uint32_t num_steps = num_train_data_samples / BATCH_SIZE;
+    for (int i = 0; i < num_steps; ++i)
+    {
+      uint32_t cur_batch_size = std::min(BATCH_SIZE, num_train_data_samples - BATCH_SIZE * i - 1);
+      cur_batch_size = std::max(1u, cur_batch_size);
+
+      config.training_context.batch_size = cur_batch_size;
+      config.training_context.num_step = i + 1;
```
Comment (chunseoklee) [inline onert-micro/eval-driver/TrainingDriver.cpp:233]: AFAIU, num_step for ADAM optimizer should not be reset by each epoch. 
```suggestion
      config.training_context.num_step++;
```

Comment (chunseoklee) [reply onert-micro/eval-driver/TrainingDriver.cpp:233]: or 
```suggestion
      config.training_context.adam_step++;
```

Comment (BalyshevArtem) [reply onert-micro/eval-driver/TrainingDriver.cpp:233]: I see, I will change it

**Comment by BalyshevArtem** [comment]
@chunseoklee, can I start split this draft ad merging it?

**Comment by chunseoklee** [comment]
> @chunseoklee, can I start split this draft ad merging it?

Sure.

**Comment by chunseoklee** [comment]
@BalyshevArtem I am trying to make onert-micro-dev module, which implement nnfw api on https://github.com/chunseoklee/ONE/commits/v3/. After drafting this, I am going to apply this to your TizenRT internal commit.

File: onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+/*
+ *  Train single step: run forward graph (with data which was set in SetInput) ->
+ *  -> calculate error (with target data which was set in SetTarget) ->
+ *  -> run backward graph -> update optimizer state -> after batch_size steps update weights
+ *  Warning: before using trainSingleStep call: 1) importTrainModel; 2) setInput; 3) setTarget
+ */
+OMStatus OMTrainingRuntimeModule::trainSingleStep(const OMConfig &config)
+{
```
Comment (chunseoklee) [inline onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:77]:
```suggestion
{
  config.training_context.num_step++;
```
we need this for update num_step for ADAM.

Comment (BalyshevArtem) [reply onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:77]: Yes, and to save this value in checkpoints files (if needed method will call)

Comment (chunseoklee) [reply onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp:77]: I mean we need somewhere to keep this num_step value

File: onert-micro/onert-micro/include/OMConfig.h [hunk 7]
```diff
@@ -17,13 +17,86 @@
...
+  MAE,
+};
+
+/*
+ * OMTrainingContext contains training specific parameters
+ * batch_size - batch size for training (Note: 1 - default one)
+ * num_of_train_layers - number of trainable last layers (Note: 0 - all layers will be trained)
+ * optimizer - optimizer which onert-micro training will be used (Note: SGD - default one)
+ * loss - loss which onert-micro training will be used (Note: CROSS_ENTROPY - default one)
+ * lambda - used by all optimizers
```
Comment (chunseoklee) [inline onert-micro/onert-micro/include/OMConfig.h:63]: Is this learning rate ? then how about using lr or learning_rate ?

File: onert-micro/onert-micro/src/train/tests/BostonHousingTask.test.cpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+  // Some value for checking that the learning process has not become worse
+  const float golden_mae_metric = 10.f;
+};
+
+TEST_F(BostonHousingTaskTest, ADAM_MSE_P)
+{
+  // Create BostonHousing data handler
+  BostonHousingTask<float> bostonTask;
+
```
Comment (BalyshevArtem) [inline onert-micro/onert-micro/src/train/tests/BostonHousingTask.test.cpp:75]: Added tests

**Comment by BalyshevArtem** [comment]
Everything is merged, close it
